GetIntoTeachingApi/Services/PostcodeService.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Text.RegularExpressions;
+using CsvHelper;
+using GeoCoordinatePortable;
+
+namespace GetIntoTeachingApi.Services
+{
+    public class PostcodeService : IPostcodeService
+    {
+        private readonly IDictionary<string, PostcodeEntry> _postcodes;
+        private const double MetersToMiles = 0.000621371;
+
+        public PostcodeService(string fixture = "./Fixtures/ukpostcodes.csv")
+        {
+            _postcodes = new Dictionary<string, PostcodeEntry>();
+
+            if (Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") == "Development")
+                fixture = "./Fixtures/ukpostcodes.dev.csv";
+
+            LoadPostcodesFixture(fixture);
+        }
+
+        public bool IsValid(string postcode)
+        {
+            if (string.IsNullOrEmpty(postcode))
+                return false;
+
+            return _postcodes.ContainsKey(Sanitize(postcode));
+        }
+
+        public double DistanceBetween(string originPostcode, string destinationPostcode)
+        {
+            var origin = _postcodes[Sanitize(originPostcode)];
+            var destination = _postcodes[Sanitize(destinationPostcode)];
+
+            return origin.Coordinate.GetDistanceTo(destination.Coordinate) * MetersToMiles;
+        }
+
+        private void LoadPostcodesFixture(string path)
+        {
+            _postcodes.Clear();
+
+            using var reader = new StreamReader(path);
+            using var csv = new CsvReader(reader, CultureInfo.InvariantCulture);
+
+            csv.Read();

```

Comment:
does this file operation have its own error handling, or does this need a check for a problems occurred when reading?

Reply:
I don't think so; there's no recovering if the CSV is dodgy, I would want it to raise an exception